### PR TITLE
Also run unit tests with zero dimensions

### LINKED ISSUE #1538: Test and extend zero-dimensional usage
Lots of components in alpaka are instantiated for some kind of dimensionality. Vectors for sizes and extents, buffers, accelerators, views, etc. Zero-dimensional versions of these might be handy in some situation, like zero-dimensional buffers: #1535. Let's revisit all alpaka constructs and check whether they allow zero-dimensional usage and extend the unit tests to cover this.

### PR DESCRIPTION
**Pull request by bernhardmgruber**
Fixes: #1538

### CODE REVIEW

**Comment by bernhardmgruber** [comment]
There are some problems around `alpaka::Vec`, so I am waiting for #1620 to easen debugging.

File: include/alpaka/kernel/TaskKernelGpuUniformCudaHipRt.hpp [hunk 1]
```diff
@@ -96,26 +96,24 @@ namespace alpaka
...
-                        throw std::runtime_error("The CUDA/HIP accelerator supports a maximum of 3 dimensions. All "
-                                                 "work division extents of the dimensions higher 3 have to be 1!");
-                    }
-                }
+                if constexpr(TDim::value > 0)
+                    for(auto i = std::min(typename TDim::value_type{3}, TDim::value); i < TDim::value; ++i)
+                        if(vec[TDim::value - 1u - i] != 1)
+                            throw std::runtime_error(
+                                "The CUDA/HIP accelerator supports a maximum of 3 dimensions. All "
+                                "work division extents of the dimensions higher 3 have to be 1!");
```
Comment (j-stephan) [inline include/alpaka/kernel/TaskKernelGpuUniformCudaHipRt.hpp]: Please add a few `{}` here.

File: include/alpaka/vec/Vec.hpp [hunk 2]
```diff
@@ -275,8 +276,9 @@ namespace alpaka
         ALPAKA_FN_HOST_ACC friend constexpr auto operator+(Vec const& p, Vec const& q) -> Vec
         {
             Vec r;
-            for(typename TDim::value_type i = 0; i < TDim::value; ++i)
-                r[i] = p[i] + q[i];
+            if constexpr(TDim::value > 0)
+                for(typename TDim::value_type i = 0; i < TDim::value; ++i)
+                    r[i] = p[i] + q[i];
```
Comment (j-stephan) [inline include/alpaka/vec/Vec.hpp]:
```suggestion
            if constexpr(TDim::value > 0)
            {
                for(typename TDim::value_type i = 0; i < TDim::value; ++i)
                    r[i] = p[i] + q[i];
            }
```

Also for the other instances.

File: include/alpaka/vec/Vec.hpp [hunk 3]
```diff
@@ -285,8 +287,21 @@ namespace alpaka
         ALPAKA_FN_HOST_ACC friend constexpr auto operator-(Vec const& p, Vec const& q) -> Vec
         {
             Vec r;
-            for(typename TDim::value_type i = 0; i < TDim::value; ++i)
-                r[i] = p[i] - q[i];
+#if BOOST_COMP_NVCC && BOOST_COMP_NVCC < BOOST_VERSION_NUMBER(11, 3, 0)
+            if(TDim::value > 0)
```
Comment (j-stephan) [inline include/alpaka/vec/Vec.hpp:293]: Are older nvcc versions unable to consume `if constexpr`?

**Review by j-stephan — CHANGES_REQUESTED** [review]
A few nitpicks:

File: include/alpaka/kernel/TaskKernelGpuUniformCudaHipRt.hpp (hunk 1 above)
Comment (bernhardmgruber) [reply include/alpaka/kernel/TaskKernelGpuUniformCudaHipRt.hpp]: Do we have a style guide recommendation for that? I can only find it for `if` statements, where parentheses should be omitted if the body is a single statement: https://github.com/ComputationalRadiationPhysics/contributing/blob/master/codingGuideLines/cpp.md#12-if-statements

Personally, I also try to avoid all braces so I have less lines wasted with the closing parentheses. I did add them in a few places though, because of nvcc bugs with `if constexpr`.

File: include/alpaka/vec/Vec.hpp (hunk 3 above)
Comment (bernhardmgruber) [reply include/alpaka/vec/Vec.hpp:293]: Not in this specific instance, where you will get a syntax error with nvcc 11.2 and below. Generally, `if constexpr` is supported.

File: include/alpaka/kernel/TaskKernelGpuUniformCudaHipRt.hpp (hunk 1 above)
Comment (j-stephan) [reply include/alpaka/kernel/TaskKernelGpuUniformCudaHipRt.hpp]: > parentheses should be omitted if the body is a single statement:

Parentheses should be omitted if the body is a *one liner*, not a single statement. I don't think the construction above counts as one line...

Comment (bernhardmgruber) [reply include/alpaka/kernel/TaskKernelGpuUniformCudaHipRt.hpp]: I find that fragile. That would mean if I would refactor a variable name and clang-format then decides to break, that I would need to add braces as well.

**Comment by bernhardmgruber** [comment]
I added a few more unnecessary braces for you @j-stephan.

**Comment by j-stephan** [comment]
They are necessary... from a certain point of view.

File: include/alpaka/mem/buf/omp5/Set.hpp [hunk 4]
```diff
@@ -96,20 +97,23 @@ namespace alpaka
...
-                using Vec1D = Vec<Dim1D, Idx>;
-
                 static_assert(Dim<TView>::value == 0u, "The view is required to have dimensionality 0!");
                 static_assert(Dim<TExtent>::value == 0u, "The extent is required to have dimensionality 0!");
 
-                return createTaskKernel<AccOmp5<Dim1D, Idx>>(
-                    WorkDivMembers<Dim1D, Idx>(Vec1D::zeros(), Vec1D::zeros(), Vec1D::zeros()),
-                    MemSetKernel(),
+                using ExtIdx = Idx<TExtent>;
+                using Dim0 = DimInt<0u>;
```
Comment (psychocoderHPC) [inline include/alpaka/mem/buf/omp5/Set.hpp]: Why do you change this kernel? This kernel must be a one-dimensional kernel to emulate the memset function on the device.

File: include/alpaka/mem/buf/omp5/Set.hpp [hunk 5]
```diff
@@ -96,20 +97,23 @@ namespace alpaka
...
+                using Acc = AccOmp5<Dim0, ExtIdx>;
+                auto kernel = [] ALPAKA_FN_ACC(const Acc&, std::uint8_t b, std::uint8_t* mem)
+                {
+                    for(auto i = 0u; i < sizeof(Elem<TView>); i++)
```
Comment (psychocoderHPC) [inline include/alpaka/mem/buf/omp5/Set.hpp]: Please call the original kernel `MemSetKernel` because your for loop is not covering pitched memory.

File: include/alpaka/mem/buf/oacc/Set.hpp [hunk 6]
```diff
@@ -98,20 +98,23 @@ namespace alpaka
...
                 static_assert(Dim<TView>::value == 0u, "The view is required to have dimensionality 0!");
                 static_assert(Dim<TExtent>::value == 0u, "The extent is required to have dimensionality 0!");
 
-                return createTaskKernel<AccOacc<DimInt<1u>, Idx>>(
-                    WorkDivMembers<DimInt<1u>, Idx>(Vec1D::zeros(), Vec1D::zeros(), Vec1D::zeros()),
-                    MemSetKernel(),
+                using ExtIdx = Idx<TExtent>;
+                using Dim0 = DimInt<0u>;
+                using Acc = AccOacc<Dim0, ExtIdx>;
+                auto kernel = [] ALPAKA_FN_ACC(const Acc&, std::uint8_t b, std::uint8_t* mem)
```
Comment (psychocoderHPC) [inline include/alpaka/mem/buf/oacc/Set.hpp]: Same as in the Omp5 `Set` function. Please keep the old implementation.

**Review by psychocoderHPC — CHANGES_REQUESTED** [review]

File: include/alpaka/mem/buf/omp5/Set.hpp (hunk 4 above)
Comment (bernhardmgruber) [reply include/alpaka/mem/buf/omp5/Set.hpp]: AFAIK I only changed the implementation of the zero dimensional Set implementation, where there are no pitches and we only set `sizeof(TElem)` bytes. I can add a comment to point that out more explicitely.

File: include/alpaka/kernel/TaskKernelGpuUniformCudaHipRt.hpp [hunk 7]
```diff
@@ -96,12 +96,16 @@ namespace alpaka
...
+                    for(auto i = std::min(typename TDim::value_type{3}, TDim::value); i < TDim::value; ++i)
                     {
-                        throw std::runtime_error("The CUDA/HIP accelerator supports a maximum of 3 dimensions. All "
-                                                 "work division extents of the dimensions higher 3 have to be 1!");
+                        if(vec[TDim::value - 1u - i] != 1)
+                        {
+                            throw std::runtime_error(
+                                "The CUDA/HIP accelerator supports a maximum of 3 dimensions. All "
+                                "work division extents of the dimensions higher 3 have to be 1!");
+                        }
```
Comment (j-stephan) [inline include/alpaka/kernel/TaskKernelGpuUniformCudaHipRt.hpp:108]: It is perfect! Perfect in every way!

**Review by j-stephan — APPROVED** [review]

File: include/alpaka/kernel/TaskKernelGpuUniformCudaHipRt.hpp (hunk 7 above)
Comment (bernhardmgruber) [reply include/alpaka/kernel/TaskKernelGpuUniformCudaHipRt.hpp:108]: :*

**Review by psychocoderHPC — APPROVED** [review]
